### PR TITLE
Fix Inductor scheduler reordering random ops (#181136)

### LINKED ISSUE #181136: Inductor should respect rand op order in eager function
The following code fails with compile. In eager code, rng order is `a -> b -> c`. In compiled code, rng order is `b -> c -> a`.

```python
import torch

def fn():
    a = torch.randn(2, 4, 4, 4, device="cuda")
    b = torch.randn(2, 3, 4, 4, device="cuda")
    c = torch.randn(4, 3, 3, 3, device="cuda")
    y = torch.nn.functional.conv2d(b, c, padding=1)
    z = a + y
    return z

f_c = torch.compile(fn)

torch.manual_seed(42)
eager_out = fn()

torch.manual_seed(42)
eager_out1 = fn()

torch.manual_seed(42)
compile_out = f_c()

torch.testing.assert_close(eager_out, eager_out1)   # PASS
torch.testing.assert_close(compile_out, eager_out)    # FAIL
```

cc @pbelevich @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by liqiangxl**
Fix #181136 when `fallback_random = True`

The Inductor scheduler reorders `torch.randn` calls based on data dependencies, causing them to consume global RNG state out of program order. This produces silent numerical mismatches vs eager mode.

Add `_add_random_dep_chain()` to chain nodes with the `nondeterministic_seeded tag via WeakDep`, preserving program order.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
@liqiangxl can you fix that lint failure?

**Comment by liqiangxl** [comment]
> @liqiangxl can you fix that lint failure?

Fixed.

**Review by eellison — COMMENTED** [review]
you might consider doing this at the fx level with effects or control deps so we dont reorder prior to inductor

**Comment by liqiangxl** [comment]
> you might consider doing this at the fx level with effects or control deps so we dont reorder prior to inductor

Good point. Moved from `scheduler.py` to `fx_passes/post_grad.py`. 
I also need to add extra changes to `lowering.py, control_deps_op_lowering` to flatten tuples, e.g. when RNG ops such as `aten.native_dropout` has `tuple-output `.  

Note that this does change `control_deps` lowering behavior for nested dependency values, but I believe it changes it in the intended direction. Previously, if an entry in `additional_deps` was a tuple/list output, lowering saw only the outer container. Since the container is not an `IRNode`, the dependency was silently ignored and no scheduler ordering edge was recorded. That does not seem like intentional semantics for `control_deps`; it means the FX graph appears to express an ordering constraint, but the constraint disappears during lowering.


**Here is AI generated more details:**

One important edge case is tuple-output RNG ops, such as `aten.native_dropout`,
`_fused_dropout`, and some attention/dropout variants. These return multiple
tensors, for example:

```python
dropped, mask = aten.native_dropout(...)
```

At FX level, the dependency is the single `native_dropout` node. But during
lowering, that FX node evaluates to a nested Python value, roughly:

```python
(TensorBox(dropped), TensorBox(mask))
```

If lowering iterates over `additional_deps` directly, it sees the outer tuple
rather than the tensor IR nodes inside it. Since the tuple itself is not an
`IRNode`, the dependency is skipped and no scheduler ordering edge is recorded.

To handle this, `control_deps` lowering now flattens dependency values with
`pytree.tree_leaves(additional_deps)` before filtering for `IRNode`s:

```python
for dep in pytree.tree_leaves(additional_deps):
    if isinstance(dep, IRNode):
        dep.realize()
        dep_names.append(dep.get_name())
```

This makes tuple/list/dict outputs work naturally: for `native_dropout`, the
`dropped` and `mask` tensor leaves are visible to lowering, so real buffer deps
are recorded for the scheduler.

The regression test covers this with:

```python
dropped, _ = aten.native_dropout(...)
b = torch.randn(...)
c = torch.randn(...)
y = conv2d(b, c)
return dropped + y
```

Without flattening tuple-output dependencies, compiled output can diverge from
eager because `randn` may be scheduled before `native_dropout`, consuming CUDA
RNG in a different order. With the fix, the ordering edge is preserved and
eager/compiled match.

**Comment by liqiangxl** [comment]
> @pytorchbot merge

Fixed CI captured test failure in `python test/inductor/test_control_deps.py TestControlDeps.test_control_deps_orders_void_op_across_nested_calls`
